kafka-client/src/main/java/dev/responsive/kafka/internal/db/rs3/client/grpc/GrpcMessageQueue.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2025 Responsive Computing, Inc.
+ *
+ * This source code is licensed under the Responsive Business Source License Agreement v1.0
+ * available at:
+ *
+ * https://www.responsive.dev/legal/responsive-bsl-10
+ *
+ * This software requires a valid Commercial License Key for production use. Trial and commercial
+ * licenses can be obtained at https://www.responsive.dev
+ */
+
+package dev.responsive.kafka.internal.db.rs3.client.grpc;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class GrpcMessageQueue<T> {
+  private final ConcurrentLinkedQueue<T> queue = new ConcurrentLinkedQueue<>();
```

Comment:
why not just use a [BlockingQueue](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html) to implement this?

Reply:
Probably a premature optimization. I think blocking queues require a lock on every queue access. I had hoped to avoid locking when the queue was not empty.